### DISCUSSION TITLE
Versioning Improvments

### DISCUSSION THREAD

**Discussion by james-d-elliott** [thuliteio/doks#621] [Q&A]
So as per a previous discussion it would be nice if versioning determined all versions available from the dropdown. I believe I've managed to achieve this with the following snippet in header.html:

```html
    {{ if and (eq .Site.Params.options.docsVersioning true) (in .Site.Params.Sections.Versioning .Section) -}}
    {{ $selected := .Site.Params.docsVersion }}
    {{ $url := .RelPermalink }}
    {{ range $release := sort (index $.Site.Data "versions") "group" "desc" -}}
    {{ $versions := sort $release.versions "v" "desc" -}}
    {{ range $i, $version := $versions -}}
    {{ $prefix := (printf "/%s/%s/" $current.Section $version.v) }}
    {{ if hasPrefix $url $prefix }}
    {{ $selected = $version.v }}
    {{ end }}
    {{ end }}
    {{ end }}

    <div class="dropdown ms-3">
      <button class="btn btn-doks-light dropdown-toggle" id="doks-versions" data-bs-toggle="dropdown" aria-expanded="false" data-bs-display="static" aria-label="Toggle version menu">
        <span class="d-none d-lg-inline">Authelia</span> v{{ $selected }}
      </button>
      <ul class="dropdown-menu dropdown-menu-end shadow rounded border-0" aria-labelledby="doks-versions">
        <li><a class="dropdown-item{{ if eq $selected .Site.Params.docsVersion }} current{{ end }}"{{ if eq $selected .Site.Params.docsVersion }} aria-current="true"{{ end }} href="/{{ $current.Section }}/{{ .Site.Params.docsVersion }}/prologue/introduction/">Latest ({{ .Site.Params.docsVersion }}.x)</a></li>
        <li><hr class="dropdown-divider"></li>
        {{ range $release := sort (index $.Site.Data "versions") "group" "desc" -}}
        {{ $versions := sort $release.versions "v" "desc" -}}
        {{ range $i, $version := $versions -}}
        {{ if not (eq $version.v $.Site.Params.docsVersion) -}}
        <li><a class="dropdown-item{{ if eq $selected $version.v }} current{{ end }}"{{ if eq $selected $version.v }} aria-current="true"{{ end }} href="/{{ $current.Section }}/{{ $version.v }}/prologue/introduction/">v{{ $version.v }}</a></li>
        {{ end }}
        {{ end }}
        {{ end }}
        <li><hr class="dropdown-divider"></li>
        <li><a class="dropdown-item" href="/{{ $current.Section }}/versions/">All versions</a></li>
      </ul>
    </div>
    {{ end -}}
 ```
 
 Is this something I should PR? Does it look reasonable?

**Comment by h-enk** [comment]
Nice one! It certainly does look reasonable, but I'd have to test this out /see it working for myself. Your PR would surely be appreciated!

**Reply by james-d-elliott** [reply]
https://authelia-staging.netlify.app/configuration/4.35/prologue/introduction/

It's still in heavy development https://github.com/authelia/doks

**Reply by james-d-elliott** [reply]
This looks strange in dark mode: https://authelia-staging.netlify.app/configuration/versions/

It's using some -primary classes which nothing else uses, not entirely sure about css.

Edit: ~~Also curiously versioned pages under content/configuration/4.x/ do not use layouts/configuration/single.html they use layouts/_default/single.html though this is probably easy to solve I just don't know how.~~ Something strange is occuring here, will try and narrow it down.

**Reply by james-d-elliott** [reply]
I think I've just worked around an issue you had in a hacky way, it probably isn't the way it was intended actually.

**Reply by james-d-elliott** [reply]
Yeah so the issue with the way I've done it is I've used subfolders, but I think you used branches based on the docs? I think all of my issues are related to doing it that way instead of via version branches. Realistically if we can build all pages with the data file from master, my method *should* still work technically.

**Reply by h-enk** [reply]
> Yeah so the issue with the way I've done it is I've used subfolders, but I think you used branches based on the docs?

Yeah, there are a number of ways to do this (subfolders being one of them). I set it up to use branches, as I recall it, this was the most solid way, to me then at least. I based it off the way Bootstrap set it up — see e.g.  https://github.com/twbs/bootstrap/tree/main/site/content/docs — but not quite, they also use a [data file](https://github.com/twbs/bootstrap/blob/main/site/data/docs-versions.yml) to generate more stuff. 

>  Realistically if we can build all pages with the data file from master, my method should still work technically.

Hmm. interesting. If you have a solid, generic way of doing that, that would possibly be something to accommodate for in Doks.

And, I recall the dark mode issues (I left them then, to go forward). To be fixed (low complexity).

**Reply by james-d-elliott** [reply]
> Hmm. interesting. If you have a solid, generic way of doing that, that would possibly be something to accommodate for in Doks.

Yep, that's going to be the issue.. I plan on deploying our docs rework without versioning and work on it. If I manage to get it worked out I'll try to throw together a PR, I know this has to be possible in an intuitive way.

> And, I recall the dark mode issues (I left them then, to go forward). To be fixed (low complexity).

Yep, I figured this was the case. I am not too worried, I am more interested in getting the technical side running smoothly.
